src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,272 @@
+      this.executionBlockingStubSupplier = executionBlockingStubSupplier;
+    }
+
+    ExecuteResponse start() throws IOException, InterruptedException {
+      // Execute has two components: the Execute call and (optionally) the WaitExecution call.
+      // This is the simple flow without any errors:
+      //
+      // - A call to Execute returns streamed updates on an Operation object.
+      // - We wait until the Operation is finished.
+      //
+      // Error possibilities:
+      // - An Execute call may fail with a retriable error (raise a StatusRuntimeException).
+      //   - If the failure occurred before the first Operation is returned, we retry the call.
+      //   - Otherwise, we call WaitExecution on the Operation.
+      // - A WaitExecution call may fail with a retriable error (raise a StatusRuntimeException).
+      //   In that case, we retry the WaitExecution call on the same operation object.
+      // - A WaitExecution call may fail with a NOT_FOUND error (raise a StatusRuntimeException).
+      //   That means the Operation was lost on the server, and we will retry to Execute.
+      // - Any call can return an Operation object with an error status in the result. Such
+      //   Operations are completed and failed; however, some of these errors may be retriable.
+      //   These errors should trigger a retry of the Execute call, resulting in a new Operation.
+      Preconditions.checkState(operation == null);
+
+      ExecuteResponse response = null;
+      while (response == null) {
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                backoff);
+
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              backoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(operation == null);
+
+      try {
+        Iterator<Operation> operations = executionBlockingStubSupplier.get().execute(request);
+        return handleStreamOperations(operations, /* resetBackoff */ false);
+      } catch (StatusRuntimeException e) {
+        if (operation != null) {
+          // By returning null, we are going to call WaitExecution in a loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(operation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(operation.getName())
+          .build();
+      try {
+        Iterator<Operation> operations = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+        return handleStreamOperations(operations, /* resetBackoff */ true);
```

Comment:
By reseting backoff, we tell retrier keep retrying in case of an error.

We don't want to reset backoff for Execute() since if there is an error:
1. It happens before we received a first response, we want to ensure the retry counter is increased and call Execute() again.
2. Otherwise, we will fallback to WaitExecution() loop.

However, we want to reset backoff for WaitExecution() so we can "infinitely" wait for the execution to complete as long as they are making progress (by returning response with WaitExecution()).